src/main/java/net/neoforged/neoforge/network/connection/ConnectionPhase.java

```diff
@@ -0,0 +1,51 @@
+package net.neoforged.neoforge.network.connection;
+
+import net.minecraft.client.multiplayer.ClientPacketListener;
+import net.minecraft.network.PacketListener;
+import net.minecraft.network.protocol.configuration.ClientConfigurationPacketListener;
+import net.minecraft.network.protocol.configuration.ServerConfigurationPacketListener;
+import net.minecraft.network.protocol.game.ClientGamePacketListener;
+import net.minecraft.network.protocol.game.ServerGamePacketListener;
+import net.minecraft.network.protocol.game.ServerPacketListener;
+
+public enum ConnectionPhase {
+    COMMON,
+    CONFIGURATION,
+    PLAY;
+
+    public boolean isConfiguration() {
+        return this == COMMON || this == CONFIGURATION;
+    }
+
+    public boolean isPlay() {
+        return this == COMMON || this == PLAY;
+    }
+
+    public static ConnectionPhase fromPacketListener(final PacketListener packetListener) {
+        if (packetListener instanceof ClientConfigurationPacketListener) {
+            return ConnectionPhase.CONFIGURATION;
+        }
+
+        if (packetListener instanceof ClientGamePacketListener) {
+            return ConnectionPhase.PLAY;
```

Comment:
As a general comment, do we want to use `GAME` instead? I think "play" is yarn. We can keep play internally cause the diff might be too large otherwise, but for modder-facing APIs it would probably be better to use "game". What do you think?

Reply:
It is actually called `PLAY` in ConnectionProtocol